### ISSUE TITLE
React Error when changing type of DatePickerInput

### ISSUE THREAD

**Issue by aviemet** [mantinedev/mantine#4650]
### What package has an issue

@mantine/dates

### Describe the bug

When changing the type of a controlled DatePickerInput from 'default' to 'ranged', React throws an error about state update timing:

`Warning: Cannot update a component ('ComponentWithLocalState') while rendering a different component ('@mantine/dates/DatePickerInput'). To locate the bad setState() call inside '@mantine/dates/DatePickerInput', follow the stack trace as described in https://reactjs.org/link/setstate-in-render`

It appears to be an issue with the call to `_setValue` on [this line in `use-uncontrolled-dates`](https://github.com/mantinedev/mantine/blob/33f51e1a7e755ef189d3e87ce17bd9d78a8a0742/src/mantine-dates/src/hooks/use-uncontrolled-dates/use-uncontrolled-dates.ts#L38). It seems that calling state setters inside the render body of a component has high potential to cause issues, so the React team introduced a specific error calling out this pattern. If that state call is moved inside of a `useEffect`, it should eliminate the error.  I'd be happy to do a PR.

### What version of @mantine/hooks page do you have in package.json?

6.0.19

### If possible, please include a link to a codesandbox with the reproduced problem

https://codesandbox.io/s/serene-mclean-hzhfpf?file=/src/DateInputErrorExample.tsx

### Do you know how to fix the issue

Yes

### Are you willing to participate in fixing this issue and create a pull request with the fix

Yes

### Possible fix

I did some local testing, and moving the bulk of the component's function body into a useEffect has solved the issue:

```typescript

export function useUncontrolledDates<Type extends DatePickerType = 'default'>({
  type,
  value,
  defaultValue,
  onChange,
}: UseUncontrolledDates<Type>) {
  const [_value, _setValue] = useUncontrolled<any>({
    value,
    defaultValue,
    onChange,
    finalValue: getEmptyValue(type),
  });

  useEffect(() => {
    // Type has changed. Do some checks or resets
    if (value === undefined) {
      const _finalValue = defaultValue !== undefined ? defaultValue : getEmptyValue(type);
      _setValue(_finalValue);
    } else if (process.env.NODE_ENV === 'development') {
      // Throw errors in dev mode in case type of value isn't correct
      switch (type) {
        case 'default':
          if (value !== null && typeof value !== 'string') {
            // eslint-disable-next-line no-console
            console.error(
              '[@mantine/dates/use-uncontrolled-dates] Value must be type of `null` or `string`'
            );
          }
          break;
        case 'multiple':
          if (!(value instanceof Array)) {
            // eslint-disable-next-line no-console
            console.error(
              '[@mantine/dates/use-uncontrolled-dates] Value must be type of `string[]`'
            );
          }
          break;
        case 'range':
          if (!(value instanceof Array) || value.length !== 2) {
            // eslint-disable-next-line no-console
            console.error(
              '[@mantine/dates/use-uncontrolled-dates] Value must be type of `[string, string]`'
            );
          }
          break;
      }
    }
  }, [type]);

  return [_value ?? defaultValue ?? getEmptyValue(type), _setValue];
}
```

**Comment by rtivital** [comment]
You need to update controlled value on your side – https://codesandbox.io/s/little-frog-g4f2gm

**Comment by aviemet** [comment]
I appreciate the quick fix suggestion, but my use case is a bit more complicated than the simple example I presented in the codesandbox and I'm still having issues with this component. I'm sorry to reopen this, but I really do think the hook deserves a slight reworking, either that or I could really benefit from some more guidance.

After posting this, I found [the discussion where  `useUncontrolledDates` was modified](https://github.com/mantinedev/mantine/issues/3919). I can see the logic in passing this responsibility off to the developer, especially as there are two choices; either to nullify or preserve an input value when the type changes. I decided in my use case that I wanted to preserve the input value if the type changed, so I set about implementing this behavior in a proxy component for my project.

Doing so surfaced another set of errors because I can't directly pass the `Date` value from the input back into the input. It triggers one of the warnings in the `useUncontrolledDates` hook saying that the value must of type string. I find this confusing because none of the valid types for the value are type string, the signature is `DateValue | Date[] | DatesRangeValue`. 

In comparing types on the input values, I've gotten even more confused because the signature for the value prop on `DatePickerInputProps` is different than that of the value prop in `DatePickerInput`, and the `value` parameter for its `onChange` prop. None of them allow type `string`, but that is the only type checked for in the `useUncontrolledDates` hook. I'm clearly a bit confused, and left wondering what I'm missing.

[I created another example to show these issues](https://codesandbox.io/s/serene-mclean-hzhfpf?file=/src/AnotherExample/DateInput.tsx)

My thought is still that the hook should handle the type change case to prevent the React error, or at least catch it and provide a useful warning to the developer indicating what the problem is. I understand there was already a discussion about this, and it led to the current implementation, so I'm happy to be wrong about that, but that's my initial take. I would also really like to see these quirks discussed in the documentation. If any changes are warranted, I'm happy to take a stab at them and do a PR.

Anyway, thank you for an amazing UI library, and I really appreciate the help with this issue.
